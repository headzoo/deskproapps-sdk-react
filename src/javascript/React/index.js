import React from 'react';
import ReactDOM from 'react-dom';
import AbstractApp from './AbstractApp';
import * as Core from '../Core';

function instance(UserApp) {
  return class extends AbstractApp {
    render = () => this.renderApp(UserApp)
  };
}

function render(DeskproApp, domselector) {
  return Core.dpapp.onMount().then(() => {
    ReactDOM.render(<DeskproApp dpapp={Core.dpapp} name="TRELLO" />, document.querySelector(domselector));
    return true;
  });
}

function connect(app) {
  const newApp = instance(app);
  return { render: domselector => render(newApp, domselector) };
}

export { instance, render };
export default connect;

