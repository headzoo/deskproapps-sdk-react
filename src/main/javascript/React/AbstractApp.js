import React, { PropTypes } from 'react';
import { Container, Header, Segment, Icon, Menu, Loader } from 'semantic-ui-react/src';
import { ChildApp } from 'deskproapps-sdk-core';

import * as UICommand from '../UICommand';

function isObject(item) {
  return (item && typeof item === 'object' && !Array.isArray(item) && item !== null);
}

function mergeDeep(target, source) {
  if (isObject(target) && isObject(source)) {
    Object.keys(source).forEach(key => {
      if (isObject(source[key])) {
        if (!target[key]) Object.assign(target, { [key]: {} });
        mergeDeep(target[key], source[key]);
      } else {
        Object.assign(target, { [key]: source[key] });
      }
    });
  }
  return target;
}

function mergeState(newState, state) {
  const newStateKeys = Object.keys(newState);
  if (newStateKeys.length === 0) {
    return {};
  }

  // extract only the keys present in the new state object
  const target = {};
  newStateKeys.forEach(key => {
    if (Object.prototype.hasOwnProperty.call(state, key)) {
      target[key] = state[key];
    }
  });
  return mergeDeep(target, newState);
}

class AbstractApp extends React.Component {
  static propTypes = {
    dpapp: PropTypes.object
    , name: PropTypes.string.isRequired
  };

  constructor(props) {
    super(props);

    this.initState();
    this.childapp = new ChildApp();
  }

  initState = () => {
    this.state = {
      layout: 'full',
      options: {
        visible: true
      },
      loader: {
        visible: true
      }
    };
  };

  onUICommand = (commandName, options) => {
    const { childapp } = this;
    const command = UICommand.fromString(commandName);
    const state = command !== null ? command.handle(childapp, options) : state;

    const newState = mergeState(state, this.state);
    this.setState(newState);
  };

  onCollapse = () => {
    const { layout } = this.state;
    const newLayout = layout === 'collapsed' ? 'full' : 'collapsed';

    this.setState({ layout: newLayout });
    this.childapp.onCollapse();
  };

  renderOptions = () => {

    const { layout, options } = this.state;

    if (!options.visible) {
      return (
        <Menu.Menu position="right" />
      );
    }

    if (layout === 'collapsed') {
      return (
        <Menu.Menu position="right" >
          <Menu.Item name="collapse" active={false} fitted onClick={this.onCollapse}>
            <Icon name="caret up" />
          </Menu.Item>
        </Menu.Menu>
      );
    }

    return (
      <Menu.Menu position="right" >
        <Menu.Item name="refresh" active={false} fitted onClick={this.childapp.onRefresh}>
          <Icon name="refresh" />
        </Menu.Item>

        <Menu.Item name="settings" active={false} fitted onClick={this.childapp.onSettings}>
          <Icon name="setting" />
        </Menu.Item>

        <Menu.Item name="collapse" active={false} fitted onClick={this.onCollapse}>
          <Icon name="caret up" />
        </Menu.Item>
      </Menu.Menu>
    );
  };

  renderAppHeader = (name) => {
    return (
      <Menu.Item>
        <Header size="small">
          <Icon name="cubes" />
          <Header.Content> {name} </Header.Content>
        </Header>
      </Menu.Item>
    );
  };

  renderAppContent = (Content, dpapp, props) => {

    const { layout, loader } = this.state;
    let conteStyle;
    const ui = UICommand.commandChain(this.onUICommand);

    if (layout === 'collapsed' || loader.visible) {
      conteStyle = { visibility: 'hidden' };
    }

    return (
      <Segment attached="bottom">
        <Loader active={loader.visible} inline="centered" />
        <div style={conteStyle}>
          <Content dpapp={dpapp} ui={ui} {...props} />
        </div>
      </Segment>
    );
  };

  renderApp = (App) => {
    const { dpapp, name, ...passThroughProps } = this.props;

    return (
      <Container>

        <Menu size="huge" borderless attached="top">
          { this.renderAppHeader(name) }
          { this.renderOptions() }
        </Menu>

        { this.renderAppContent(App, dpapp, passThroughProps) }

      </Container>
    );
  };

  render() {
    throw new Error('render must be implemented in a subclass');
  }
}

export default AbstractApp;
