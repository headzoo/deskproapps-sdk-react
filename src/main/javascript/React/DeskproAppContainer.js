import React, { PropTypes } from 'react';
import { Container, Header, Segment, Icon, Menu, Loader, Image } from 'semantic-ui-react/src';

import { AppEvents, UIEvents, UIConstants } from '@deskproapps/deskproapps-sdk-core';

class DeskproAppContainer extends React.Component {
  static propTypes = {
    app: PropTypes.object.isRequired
    , name: PropTypes.string.isRequired
    , mainComponent: PropTypes.object.isRequired
  };

  constructor(props) {
    super(props);
    this.registerListeners();
  }

  registerListeners = () =>
  {
    const { app } = this.props;

    const forceUpdateTrigger = () => this.forceUpdate();

    app.on(AppEvents.EVENT_EXPAND, forceUpdateTrigger);
    app.on(AppEvents.EVENT_COLLAPSE, forceUpdateTrigger);
    app.on(UIEvents.EVENT_STATE_TRANSITION, forceUpdateTrigger);
    app.on(UIEvents.EVENT_MENU_STATE_TRANSITION, forceUpdateTrigger);
    app.on(UIEvents.EVENT_BADGE_COUNTCHANGED, forceUpdateTrigger);
    app.on(UIEvents.EVENT_BADGE_VISIBILITYCHANGED, forceUpdateTrigger);
  };

  renderOptions = () => {
    const { visibility, ui } = this.props.app;

    if (ui.menu === 'hidden') {
      return ( <Menu.Menu position="right" /> );
    }

    if (visibility === 'collapsed') {
      return this.renderOptionsWhenCollapsed();
    }
    return this.renderOptionsWhenExpanded();
  };

  renderOptionsWhenCollapsed =() => {
    const { app } = this.props;
    const collapseListener = () => app.visibility === 'collapsed' ? app.expand() : app.collapse();

    return (
      <Menu.Menu position="right" >
        <Menu.Item name="collapse" active={false} fitted onClick={collapseListener} className={"app-option"}>
          <Icon name="caret down" size="large"/>
        </Menu.Item>
      </Menu.Menu>
    );
  };

  renderOptionsWhenExpanded =() => {
    const { app } = this.props;
    const collapseListener = () => app.visibility === 'collapsed' ? app.expand() : app.collapse();

    const options = [
      <Menu.Item name="refresh" active={false} fitted onClick={app.refresh} className={"app-option"}>
        <Icon name="refresh" />
      </Menu.Item>
    ];

    const hasSettings = app.settings.length > 0;
    if (hasSettings) {
      options.push(
        <Menu.Item name="settings" active={false} fitted onClick={app.ui.showSettings} className={"app-option"}>
          <Icon name="setting" />
        </Menu.Item>
      );
    }

    options.push(
      <Menu.Item name="collapse" active={false} fitted onClick={collapseListener} className={"app-option big"}>
        <Icon name="caret up" />
      </Menu.Item>
    );

    return (<Menu.Menu position="right">{ options} </Menu.Menu>);
  };

  renderAppBadge = app =>
  {
    const { badgeCount } = app.ui;
    return (
      <svg xmlns="http://www.w3.org/2000/svg" xmlnsXlink="http://www.w3.org/1999/xlink" width="18" height="18">
        <g>
          <circle cx="9" cy="9" r="9" fill="red"/>
          <text y="12" transform="translate(9)" fontSize="10px" fill="white" stroke="white">
            <tspan x="0" textAnchor="middle">{badgeCount}</tspan>
          </text>
        </g>
      </svg>
    );
  };

  renderAppHeader = (name, app) => {

    return (
      <Menu.Item className={"app-header"}>
        <Header size="tiny">
          <div style={{ position: 'relative' }} className="ui icon deskpro-app-icon">
            {
              app.ui.badge === UIConstants.VISIBILITY_VISIBLE &&
              <div style={{ position: 'absolute', top: '-9px', right: '-9px' }}>
                { this.renderAppBadge(app) }
              </div>
            }
            <img src="../assets/icon.png" style={{ width: '16px', height: '16px', border: 0}}/>
          </div>
          <Header.Content> {name.toUpperCase()} </Header.Content>
        </Header>
      </Menu.Item>
    );
  };

  renderAppContent = (Content, app, props) => {
    const invisibleStyle = { visibility: 'hidden', display: 'none' };

    const { visibility, ui } = app;
    const contentStyle = ui.state === 'loading' || visibility === 'collapsed'  ? invisibleStyle : {};

    return (
      <div style={contentStyle} id="dp-app-container">
        <Segment attached="bottom" className={"app-body"}>
            <Content dpapp={ app } />
        </Segment>
      </div>
    );
  };

  renderAppLoader = () => {
    const invisibleStyle = { visibility: 'hidden', display: 'none' };

    const { app } = this.props;
    const contentStyle = app.ui.state === 'loading' ? {} : invisibleStyle;

      return (
        <div style={contentStyle} id="dp-app-loader">
          <Segment attached="bottom" className={"app-body"}>
            <Loader active={true} inline="centered" />
          </Segment>
        </div>
      );
  };

  render = () => {
    const { app, mainComponent, name, ...passThroughProps } = this.props;

    return (
      <Container>
        <Menu borderless attached="top" className={"app-menu"}>
          { this.renderAppHeader(name, app) }
          { this.renderOptions() }
        </Menu>
        { this.renderAppLoader() }
        { this.renderAppContent(mainComponent, app, passThroughProps) }
      </Container>
    );
  };

}

export default DeskproAppContainer;
