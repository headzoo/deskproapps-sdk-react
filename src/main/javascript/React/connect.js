import React from 'react';
import ReactDOM from 'react-dom';

import AbstractApp from './AbstractApp';
import { connect as dpConnect } from '@deskproapps/deskproapps-sdk-core';


const instance = (UserApp) => {
  return class extends AbstractApp {
    render = () => this.renderApp(UserApp)
  };
};

const render = (DeskproApp, name, domselector) => {
    return dpConnect().then((dpapp) => {
        ReactDOM.render(<DeskproApp dpapp={dpapp} name={name} />, document.querySelector(domselector));
        return true;
    });
};

const connect = (app, appName) => {
    const newApp = instance(app);
    return { render: domselector => render(newApp, appName, domselector) };
};

export default connect;

