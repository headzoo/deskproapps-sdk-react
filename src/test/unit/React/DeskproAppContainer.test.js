// Link.react.js
import React from 'react';
import { createAppFromProps } from '@deskproapps/deskproapps-sdk-core';
import renderer from 'react-test-renderer';

import DeskproAppContainer from '../../../main/javascript/React/DeskproAppContainer'

class TestApp extends React.Component
{
  render() { return (<div>Hello World</div>); }
}

test('successfully render an application in initial state', done => {

  const contextProps = {
    // context
    type: 'ticket',
    entityId: '1',
    locationId: 'ticket-sidebar',
    tabId: 'tab-id',
    tabUrl: 'http://127.0.0.1'
  };

  const instanceProps = {
    appId: '1',
    appTitle: 'test',
    appPackageName: 'com.deskpro.test',
    instanceId: '1'
  };

  const dpapp = createAppFromProps({ contextProps, instanceProps });

  const component = renderer.create(
    <DeskproAppContainer app={dpapp} name="Test App" mainComponent={TestApp} />
  );

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  done();
});

test('successfully render an application with the badge count visible', done => {

  const contextProps = {
    // context
    type: 'ticket',
    entityId: '1',
    locationId: 'ticket-sidebar',
    tabId: 'tab-id',
    tabUrl: 'http://127.0.0.1'
  };

  const instanceProps = {
    appId: '1',
    appTitle: 'test',
    appPackageName: 'com.deskpro.test',
    instanceId: '1'
  };

  const dpapp = createAppFromProps({ contextProps, instanceProps });

  dpapp.ui.showBadgeCount();

  const component = renderer.create(
    <DeskproAppContainer app={dpapp} name="Test App" mainComponent={TestApp} />
  );

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  done();
});