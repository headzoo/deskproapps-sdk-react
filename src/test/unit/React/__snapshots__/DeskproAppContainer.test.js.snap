// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`successfully render an application 1`] = `
<div
  className="ui container"
>
  <div
    className="ui borderless top attached app-menu menu"
  >
    <div
      className="item app-header"
      onClick={[Function]}
    >
      <div
        className="ui tiny header"
      >
        <div
          className="ui icon deskpro-app-icon"
          style={
            Object {
              "position": "relative",
            }
          }
        >
          <img
            src="../assets/icon.png"
            style={
              Object {
                "border": 0,
                "height": "16px",
                "width": "16px",
              }
            }
          />
        </div>
        <div
          className="content"
        >
           
          TEST APP
           
        </div>
      </div>
    </div>
    <div
      className="right menu"
    >
      <a
        className="fitted item app-option"
        onClick={[Function]}
      >
        <i
          aria-hidden="true"
          className="refresh icon"
        />
      </a>
      <a
        className="fitted item app-option big"
        onClick={[Function]}
      >
        <i
          aria-hidden="true"
          className="caret up icon"
        />
      </a>
       
    </div>
  </div>
  <div
    id="dp-app-loader"
    style={
      Object {
        "display": "none",
        "visibility": "hidden",
      }
    }
  >
    <div
      className="ui bottom attached segment app-body"
    >
      <div
        className="ui active centered inline loader"
      />
    </div>
  </div>
  <div
    id="dp-app-container"
    style={Object {}}
  >
    <div
      className="ui bottom attached segment app-body"
    >
      <div>
        Hello World
      </div>
    </div>
  </div>
</div>
`;
